### PR TITLE
S2U-12 5.1.1.8 Improving i18n and feedback on calculated questions

### PR DESCRIPTION
**Pull request by jesusmmp**
https://sakaiproject.atlassian.net/browse/S2U-12

This is paired with https://github.com/sakaiproject/sakai-reference/pull/186

### CODE REVIEW

**Comment by ottenhoff** [comment]
Makes sense to me. Do you want community feedback on these PRs?

**Comment by jesusmmp** [comment]
> Makes sense to me. Do you want community feedback on these PRs?

Of course! This PR will be merged on 24 on next months. 
Thanks!

**Comment by mpellicer** [comment]
@ottenhoff Codacy will always complain because the branch is not being analyzed, it's something that has to be configured in Codacy. Do we want Codacy feedback for the S2U project? I think makes sense because the sooner we fix Codacy issues the better for the future contribution to the Sakai main branch.....

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java [hunk 1]
```diff
@@ -866,17 +859,17 @@ else if(item.getTypeId().equals(TypeFacade.IMAGEMAP_QUESTION)) {
             item.setItemMetaDataSet(prepareMetaData(item, bean));
 
             // prepare feedback, only store if feedbacks are not empty
-            if ( (bean.getCorrFeedback() != null) &&
-      			  (!bean.getCorrFeedback().equals(""))) {
-            	item.setCorrectItemFeedback(stripPtags(bean.getCorrFeedback()));
+      	  	if (!StringUtils.isEmpty(bean.getCorrFeedback())) {
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java:862]: Better using 'isNotEmpty' for more readability

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java [hunk 2]
```diff
@@ -832,14 +822,17 @@ public void saveItem(ItemAuthorBean itemauthor, AssessmentBean assessmentBean) t
...
-              if ((bean.getIncorrFeedback() != null)) {
-                	updateItemFeedback(item, ItemFeedbackIfc.INCORRECT_FEEDBACK, stripPtags(bean.getIncorrFeedback()));
-              }
-              if ((bean.getGeneralFeedback() != null)) {
-                	updateItemFeedback(item, ItemFeedbackIfc.GENERAL_FEEDBACK, stripPtags(bean.getGeneralFeedback()));
-           }
+                newFeedback = stripPtags(bean.getCorrFeedback());
+                updateItemFeedback(item, ItemFeedbackIfc.CORRECT_FEEDBACK, newFeedback, newFeedback);
+          }
+          if ((bean.getIncorrFeedback() != null)) {
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java:828]: We can also use StringUtils here right?

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java [hunk 3]
```diff
@@ -832,14 +822,17 @@ public void saveItem(ItemAuthorBean itemauthor, AssessmentBean assessmentBean) t
...
+                newFeedback = stripPtags(bean.getIncorrFeedback());
+                updateItemFeedback(item, ItemFeedbackIfc.INCORRECT_FEEDBACK, newFeedback, newFeedback);
+          }
+          if ((bean.getGeneralFeedback() != null)) {
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java:832]: We can also use StringUtils here right?

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java [hunk 4]
```diff
@@ -866,17 +859,17 @@ else if(item.getTypeId().equals(TypeFacade.IMAGEMAP_QUESTION)) {
...
-      			  (!bean.getCorrFeedback().equals(""))) {
-            	item.setCorrectItemFeedback(stripPtags(bean.getCorrFeedback()));
+      	  	if (!StringUtils.isEmpty(bean.getCorrFeedback())) {
+      	  		newFeedback = stripPtags(bean.getCorrFeedback());
+      	  		item.setCorrectItemFeedback(newFeedback, newFeedback);
       	  	}
-      	  	if ( (bean.getIncorrFeedback() != null) &&
-      			  (!bean.getIncorrFeedback().equals(""))) {
-      	  		item.setInCorrectItemFeedback(stripPtags(bean.getIncorrFeedback()));
+      	  	if (!StringUtils.isEmpty(bean.getIncorrFeedback())) {
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java:866]: Better use 'isNotEmpty' for more readability

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java [hunk 5]
```diff
@@ -866,17 +859,17 @@ else if(item.getTypeId().equals(TypeFacade.IMAGEMAP_QUESTION)) {
...
+      	  		newFeedback = stripPtags(bean.getIncorrFeedback());
+      	  		item.setInCorrectItemFeedback(newFeedback, newFeedback);
       	  	}
-      	  	if ( (bean.getGeneralFeedback() != null) &&
-      			  (!bean.getGeneralFeedback().equals(""))) {
-      	  		item.setGeneralItemFeedback(stripPtags(bean.getGeneralFeedback()));
+      	  	if (!StringUtils.isEmpty(bean.getGeneralFeedback())) {
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java:870]: Better use 'isNotEmpty' for more readability

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java [hunk 6]
```diff
@@ -1259,12 +1262,17 @@ else if (itemBean.getMaxPoints()>0 && !item.getTypeId().equals(TypeIfc.MULTIPLE_
    	
     	if (itemBean.getExactPoints() >= itemBean.getMaxPoints())
     	{
-    		
     		itemBean.setFeedback(item.getCorrectItemFeedback());
+    		if (item.getTypeId().equals(TypeIfc.CALCULATED_QUESTION)) {
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java:1266]: TypeIfc.CALCULATED_QUESTION.equals(item.getTypeId())

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java [hunk 7]
```diff
@@ -1259,12 +1262,17 @@ else if (itemBean.getMaxPoints()>0 && !item.getTypeId().equals(TypeIfc.MULTIPLE_
...
+    			itemBean.setFeedbackValue(item.getCorrectItemFeedbackValue());
+    		}
     	}
     	else
     	{
     		itemBean.setFeedback(item.getInCorrectItemFeedback());
+    		if (item.getTypeId().equals(TypeIfc.CALCULATED_QUESTION)) {
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java:1273]: TypeIfc.CALCULATED_QUESTION.equals(item.getTypeId())

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java [hunk 8]
```diff
@@ -2918,13 +2932,38 @@ private String commaDelimitedCalcQuestionAnswers(ItemDataIfc item, DeliveryBean
 		  
 		  keysString = keysString.concat(answer + ", ");
 		  answerSequence++;
-	  }
+	}
```
Comment (mpellicer) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java:2935]: Identation seems odd here

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/GradingService.java [hunk 9]
```diff
@@ -2929,6 +2936,24 @@ protected List<String> extractInstructionSegments(String instructions) {
       return segments;
   }
 
+  protected List<String> extractFeedbackSegments(String feedback) {
+      List<String> segments = new ArrayList<>();
+      String[] results = null;
+
+      if (!StringUtils.isEmpty(feedback)) {
+          String[] firstResults = CALCQ_DOUBLE_OPEN_BRACKET.split(feedback); // only works because all variables and calculations are already replaced
+          for (String r: firstResults) {
```
Comment (mpellicer) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/GradingService.java:2945]: Spacing.

File: samigo/samigo-qti/src/java/org/sakaiproject/tool/assessment/qti/helper/ExtractionHelper.java [hunk 10]
```diff
@@ -1787,43 +1787,47 @@ else if (type.endsWith("_IC")) {
...
 	  String generalItemFeedback = (String) map.get("generalItemFeedback");
+	  String updatedFeedback;
 
 	  if (generalItemFeedback==null) generalItemFeedback = "";
 	  if (TypeIfc.AUDIO_RECORDING.longValue() == typeId.longValue() ||
 			  TypeIfc.FILE_UPLOAD.longValue() == typeId.longValue() ||
 			  TypeIfc.ESSAY_QUESTION.longValue() == typeId.longValue())
 	  {
-		  if (notNullOrEmpty(incorrectItemFeedback))
+		  if (!StringUtils.isEmpty(incorrectItemFeedback))
```
Comment (mpellicer) [inline samigo/samigo-qti/src/java/org/sakaiproject/tool/assessment/qti/helper/ExtractionHelper.java:1807]: Better isNotEmpty ?

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/QuestionPoolFacadeQueries.java [hunk 11]
```diff
@@ -1250,14 +1251,14 @@ private ItemFacade getItemFacade(ItemDataIfc itemData) {
       item.setItemAttachmentSet(copyAttachment(item.getData(), itemData));
       item.setInstruction(AssessmentService.copyStringAttachment(itemData.getInstruction()));
 
-      if (itemData.getCorrectItemFeedback() != null && !itemData.getCorrectItemFeedback().equals("")) {
-    	  item.setCorrectItemFeedback(AssessmentService.copyStringAttachment(itemData.getCorrectItemFeedback()));
+      if (!StringUtils.isEmpty(itemData.getCorrectItemFeedback())) {
```
Comment (mpellicer) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/QuestionPoolFacadeQueries.java:1254]: Better use isNotEmpty

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/GradingService.java [hunk 12]
```diff
@@ -146,6 +149,8 @@
   // SAK-40942 - Error in calculated questions: the decimal representation .n or n. (where n is a number) does not work
   public static final Pattern CALCQ_FORMULA_ALLOW_POINT_NUMBER = Pattern.compile("([^\\d]|^)([\\.])([\\d])");
   public static final Pattern CALCQ_FORMULA_ALLOW_NUMBER_POINT = Pattern.compile("([\\d])([\\.])([^\\d]|$)");
+  public static final Pattern CALCQ_DOUBLE_OPEN_BRACKET = Pattern.compile(OPEN_BRACKET + OPEN_BRACKET);
+  public static final Pattern CALCQ_DOUBLE_CLOSE_BRACKET = Pattern.compile(CLOSE_BRACKET  + CLOSE_BRACKET );
```
Comment (mpellicer) [inline samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/GradingService.java:153]: Minor spacing

**Review by mpellicer — CHANGES_REQUESTED** [review]
Very very nice work @jesusmmp, you're a monster !!

I just added some code quality changes, let's re-review this again after a codacy analysis.

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java (hunk 8 above)
Comment (jesusmmp) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java:2935]: It's OK with my change.

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java (hunk 1 above)
Comment (jesusmmp) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java]: Done

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java (hunk 2 above)
Comment (jesusmmp) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java]: Done

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java (hunk 4 above)
Comment (jesusmmp) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java]: Done

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java (hunk 6 above)
Comment (jesusmmp) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java]: Done

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java (hunk 7 above)
Comment (jesusmmp) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/delivery/DeliveryActionListener.java]: Done

File: samigo/samigo-qti/src/java/org/sakaiproject/tool/assessment/qti/helper/ExtractionHelper.java (hunk 10 above)
Comment (jesusmmp) [reply samigo/samigo-qti/src/java/org/sakaiproject/tool/assessment/qti/helper/ExtractionHelper.java]: Done

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/GradingService.java (hunk 12 above)
Comment (jesusmmp) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/services/GradingService.java]: Done

File: samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/QuestionPoolFacadeQueries.java (hunk 11 above)
Comment (jesusmmp) [reply samigo/samigo-services/src/java/org/sakaiproject/tool/assessment/facade/QuestionPoolFacadeQueries.java]: Done

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java (hunk 5 above)
Comment (jesusmmp) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java]: Done

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java (hunk 3 above)
Comment (jesusmmp) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/author/ItemAddListener.java]: Done

**Comment by mpellicer** [comment]
For the conversion script, I think we can contribute the changes to https://github.com/sakaiproject/sakai-reference after the merge of https://github.com/sakaiproject/sakai-reference/pull/185

WDYT?

**Comment by mpellicer** [comment]
Thanks for the changes @jesusmmp!

Could we link the conversion script?

**Review by mpellicer — APPROVED** [review]

**Pull request merged**